05-html-css/flexbox-vs-css-grid.md

```diff
@@ -0,0 +1,153 @@
+
+## Flexbox & Grid 1-D & 2-D Layout
+
+![flexbox & Grid 1D & 2D image](imgs/grid-vs-flexbox.png)
+
+### Flexbox Lays Content Out in One Dimension
+
+All content in flexbox _flows_ horizontally or vertically inside a container.
+
+Consider the following HTML & CSS:
+
+**HTML:**
+```html
+<header>
+  <div>
+    <img src="http://via.placeholder.com/100x100" alt="placeholder">
+  </div>
+  <div>
+    <img src="http://via.placeholder.com/200x100" alt>
+  </div>
+  <div>
+    <img src="http://via.placeholder.com/300x200" alt>
+  </div>
+  <div>
+    <img src="http://via.placeholder.com/100x200" alt>
+  </div>
+  <div>
+    <img src="http://via.placeholder.com/400x300" alt>
+  </div>
+</header>
+```  
+
+**CSS:**
+```css
+header {
+  display: flex;
+  flex-direction: column;
+  flex-flow: row wrap;
+}
+img {
+  margin: .5em;
+}
+```
+Flexbox causes the images will line up in the available space left-to-right.  When the screen runs out of space elements will flow to the next line.   In this way Flexbox is _responsive_, the items wrap as needed to fit the available space.  
+
+On a wide screen:
+
+![wide-screen flexbox example](imgs/flexbox-large-width.png)
+
+On a narrower screen:
+
+![narrower-screen flexbox example](imgs/flexbox-small-width.png)
+
+You can of course cause Flexbox to lay content out in any direction, left-to-right, right-to-left, top-to-bottom, or bottom-to-top by changing the `flex-direction`.
+
+**However** you cannot align the content in 2-Dimensions.  **Flexbox is 1-Dimensional**.  Content only flows along one axis, wrapping as needed.   
+
+### Grid Lays Content Out In A... Grid
+
+Using CSS Grid we can lay the same HTML in two dimensions this way:
+
+```css
+header {
+  display: grid;
+  grid-template-columns: repeat(3, 1fr);
+}
+img {
+  margin: .5em;
+}
+body {
+  background-color: black;
+}
+```
+
+![images with grid](imgs/grid-images.png)
+
+The content is organized into both rows and columns.  The grid generates as many 3-column rows as needed to contain the content.  We see gaps where content does not fill an entire cell, and with only 5 images, the bottom row contains an empty cell.  
+
+With Flexbox the layout responds to the content, whereas with Grid the content responds to the layout.  It is often said that _Flexbox works from the flex-items out_ whereas _Grid works from the layout in_.
+
+### Flexbox Adjusts Automatically To Screen width
+
+Because Flexbox adjusts to the content the layout will respond automatically if you narrow the browser window adjusting the number of items in each row to allow them to fit.  
+
+By contrast, Grid example above will always maintain 3 columns.  There are ways to adjust the number of columns for smaller screens, by using something like: `grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));` which will generate as many columns as needed with a minimum size of 200 pixels and a maximum of 1 fraction of the row.  
+
```

Comment:
With that repeat/autofill example, a dozen students' heads just exploded. They're going to want to know a lot more about this, or we should not show the complex example.